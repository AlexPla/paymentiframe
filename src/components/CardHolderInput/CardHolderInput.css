/*
* DIV
*/
.card-holder-input__input-group {
  box-sizing: border-box;
  width: 100%;
  padding: 12px 6px 0;
}

.card-holder-input__input-group__storefront {
  margin-top: 23px;
}


/*
* INPUT FIELD
*/
.card-holder-input__input {
  font: inherit;
  font-size: 14px;
  font-weight: 300;
  margin: 0;
  box-sizing: border-box;
  display: block;
  width: 100%;
  color: #000;
  -moz-appearance: textfield;
  line-height: normal;
  box-shadow: none;
}

.card-holder-input__input__storefront {
  padding: 10px 30px;
  border: none;
  border-bottom: 1px solid #ccc;
  -webkit-appearance: none;
  border-radius: 0;
}

.card-holder-input__input__legacy,
.card-holder-input__input__webmobile {
  background: #fff;
  border: 1px solid #ccc;
  display: block;
  font-weight: 200;
  font-size: 12px;
}

.card-holder-input__input__legacy {
  outline: none;
  padding: 4px;
}

.card-holder-input__input__webmobile {
  font-size: 14px;
  padding: 12px 8px;
}

.card-holder-input__input:focus {
  outline: none;
  border-color: #000;
  transition: all .2s ease;
  -moz-transition: .2s ease all;
  -webkit-transition: all .2s ease;
}

.card-holder-input__input_invalid {
  border-color: #e55253;
}

.card-holder-input__input_invalid:focus {
  border-color: #e55253;
}

/*
* LABEL - PLACEHOLDER
*/
.card-holder-input__label__storefront {
  color: #666;
  font-size: 14px;
  font-weight: 200;
  position: absolute;
  pointer-events: none;
  left: 35px;
  top: 46px;
  -webkit-transition: all .2s ease;
  -moz-transition: .2s ease all;
  transition: all .2s ease;
}

.card-holder-input__input:focus~.card-holder-input__label__storefront {
  color: #000;
}

.card-holder-input__label__storefront__invalid {
  color: #e55253;
}

.card-holder-input__label__legacy {
  color: #000000;
  font-size: 12px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  letter-spacing: normal;
}

.card-holder-input__input__storefront:focus~.card-holder-input__label__storefront,
.card-holder-input__input__not-empty~.card-holder-input__label__storefront {
  top: 25px;
  font-size: 10px;
}


/*
* CARD ICON
*/
.card-holder-input__icon {
  fill: #666;
  position: absolute;
  top: 45px;
  left: 6px;
  cursor: default;
  text-align: center;
}

.card-holder-input__input:focus~.card-holder-input__icon {
  fill: #000;
}

.card-holder-input__icon_invalid {
  fill: #e55253;
}

.card-holder-input__input:focus~.card-holder-input__icon_invalid {
  fill: #e55253;
}


/*
* CLEAR BUTTON
*/
.card-holder-input__clear-button {
  fill: #999;
  position: absolute;
  cursor: pointer;
  display: none;
}

.card-holder-input__clear-button__storefront {
  right: 10px;
  top: 48px;
}

.card-holder-input__clear-button__legacy {
  right: 12px;
  top: 43.5px;
}

.card-holder-input__clear-button__webmobile {
  right: 17px;
  top: 53px;
}

.card-holder-input__input:focus~.card-holder-input__clear-button {
  display: block;
}


/*
* ERROR
*/
.card-holder-input__error {
  position: absolute;
  font-size: 10px;
  font-size: .714rem;
  color: #e55253;
  left: 5px;
}