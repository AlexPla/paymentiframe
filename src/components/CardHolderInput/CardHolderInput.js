import React, { Component } from 'react';
import PropTypes from 'prop-types';
import copies from '@Copies/cardHolderInput';
import './CardHolderInput.css';

class CardHolderInput extends Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      errorMessage: '',
      errorDisabled: true,
    };

    this.onInput = this.onInput.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.onClearClick = this.onClearClick.bind(this);
  }

  onInput(e) {
    const cardHolder = e.target.value;
    const errorMessage = this.validateInput(cardHolder);
    this.updateAppState(cardHolder, errorMessage);
    this.setState({ errorMessage });
  }

  onClearClick() {
    const { lang } = this.props;
    const errorMessage = copies.errors.required[lang];
    this.updateAppState('', errorMessage);
    this.setState({
      errorMessage,
      errorDisabled: false,
    });
  }

  onBlur() {
    let { value: cardHolder } = this.props;
    cardHolder = cardHolder.trim();
    const errorMessage = this.validateInput(cardHolder);
    this.updateAppState(cardHolder, errorMessage);
    this.setState({
      errorDisabled: false,
      errorMessage,
    });
  }

  validateInput(input) {
    const pattern = new RegExp(/^[a-zA-Zñçáéíóúàèòïü´ÑÇÁÉÍÓÚÀÈÒÏÜ]+([\‒\-\-]+[a-zA-Zñçáéíóúàèòïü´ÑÇÁÉÍÓÚÀÈÒÏÜ]+)*([\s]+([a-zA-Zñçáéíóúàèòïü´ÑÇÁÉÍÓÚÀÈÒÏÜ]+([\‒\-\-]+[a-zA-Zñçáéíóúàèòïü´ÑÇÁÉÍÓÚÀÈÒÏÜ]+)*))*$/); // eslint-disable-line no-useless-escape
    const { lang } = this.props;
    switch (true) {
      case input.length === 0:
        return copies.errors.required[lang];
      case input.length < 2:
        return 'systemMessages.SM18';
      case input.length > 100:
        return 'maximum length exceeded';
      case !pattern.test(input):
        return 'systemMessages.SM26';
      default:
        return '';
    }
  }

  updateAppState(cardHolder, errorMessage) {
    const { updateFields, updateErrors } = this.props;
    updateFields({ cardHolder });
    updateErrors({
      key: 'cardHolder',
      value: !!errorMessage,
    });
  }

  render() {
    const { value, lang } = this.props;
    const { errorMessage, errorDisabled } = this.state;
    const isEmpty = (value).toString().length === 0;
    return (
      <div className="card-holder-input__input-group">
        <input
          className={`card-holder-input__input ${!isEmpty && 'card-holder-input__input_not-empty'} ${!errorDisabled && errorMessage && 'card-holder-input__input_invalid'}`}
          type="text"
          autoComplete="on"
          id="cardHolderInput"
          value={value}
          onInput={this.onInput}
          onBlur={this.onBlur}
        />
        <label htmlFor="cardHolderInput" className="card-holder-input__label">{ copies.holder[lang] }</label>
        <svg className="card-holder-input__icon" version="1.1" xmlns="http://www.w3.org/2000/svg" width="20px" height="20px" viewBox="0 0 150 150" preserveAspectRatio="xMidYMid meet">
          <path d="M95.540625,75.225 C108.1875,68.296875 116.75625,55.21875 116.75625,40.21875 C116.75625,18.046875 98.025,0 75,0 C51.975,0 33.24375,18.046875 33.24375,40.21875 C33.24375,55.21875 41.8125,68.296875 54.459375,75.225 C23.071875,83.85 0,111.61875 0,144.525 C0,144.890625 0.028125,145.2375 0.084375,145.95 C0.084375,148.190625 1.89375,150 4.134375,150 L145.875,150 C148.10625,150 149.953125,147.825 149.953125,145.584375 C149.971875,145.2375 150,144.890625 150,144.525 C150,111.61875 126.928125,83.85 95.540625,75.225 L95.540625,75.225 Z M41.353125,40.21875 C41.353125,22.51875 56.446875,8.109375 75,8.109375 C93.553125,8.109375 108.646875,22.51875 108.646875,40.21875 C108.646875,57.946875 93.553125,72.365625 75,72.365625 C56.446875,72.365625 41.353125,57.946875 41.353125,40.21875 L41.353125,40.21875 Z M8.165625,141.890625 C9.609375,107.8125 39.0375,80.5125 75,80.5125 C110.840625,80.5125 140.19375,107.64375 141.796875,141.88125 L8.165625,141.88125 L8.165625,141.890625 Z" />
        </svg>
        { !isEmpty
          && (
          <svg className="card-holder-input__clear-button" onClick={this.onClearClick} version="1.1" xmlns="http://www.w3.org/2000/svg" width="15px" height="15px" viewBox="0 0 150 150" preserveAspectRatio="xMidYMid meet">
            <path d="M75.253125,1.021875 C34.10625,1.021875 0.76875,34.359375 0.76875,75.50625 C0.76875,116.653125 34.10625,150 75.253125,150 C116.41875,150 149.7375,116.6625 149.7375,75.515625 C149.7375,34.36875 116.41875,1.021875 75.253125,1.021875 L75.253125,1.021875 Z M107.85,97.715625 L97.4625,108.09375 L75.253125,85.884375 L53.053125,108.09375 L42.675,97.715625 L64.875,75.515625 L42.675,53.30625 L53.053125,42.928125 L75.2625,65.1375 L97.4625,42.928125 L107.85,53.30625 L85.640625,75.515625 L107.85,97.715625 L107.85,97.715625 Z" />
          </svg>
          )
        }
        { !errorDisabled && errorMessage
          && (
          <span className="card-holder-input__error">{` ${errorMessage} `}</span>
          )
        }
      </div>
    );
  }
}

CardHolderInput.propTypes = {
  updateFields: PropTypes.func.isRequired,
  updateErrors: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  lang: PropTypes.string.isRequired,
};

export default CardHolderInput;
