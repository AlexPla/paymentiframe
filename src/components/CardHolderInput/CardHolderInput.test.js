import React from 'react';
import { configure, mount } from 'enzyme';
import toJson from "enzyme-to-json";
import Adapter from 'enzyme-adapter-react-16';

import CardHolderInput from './CardHolderInput';

configure({ adapter: new Adapter() });

let component, input, updateFieldsMock, updateErrorsMock;

describe('Component CardHolderInput:', () => {

  beforeEach(function () {
    updateFieldsMock = jest.fn();
    updateErrorsMock = jest.fn();
    component = mount(<CardHolderInput
      lang={ 'es' }
      updateFields={ updateFieldsMock }
      updateErrors={ updateErrorsMock }
      value={ '' } />);
    input = component.find(".card-holder-input__input");
  });

  it('should mount', () => {
    expect(component).toBeDefined();
  });

  it('should match snapshot', () => {
    expect(toJson(component, { mode: "deep" })).toMatchSnapshot();
  });

  it('should call updateFields method from props and pass proper object with new field value as an argument', () => {
    input.prop('onInput')({
      target: {
        value: "changed"
      }
    });
    expect(updateFieldsMock.mock.calls[0][0]).toEqual({ card_holder: 'changed' });
  });

  it('\'should change state.errorDisabled to true on input', () => {
    component.setState({ errorMessage: 'errorMessage' });
    input.prop('onInput')({
      target: {
        value: "changed"
      }
    });
    component.update();
    expect(component.state('errorDisabled')).toBe(true);
  });

  it('should validate input on input and clear state.errorMessage for proper name', () => {
    input.prop('onInput')({
      target: {
        value: "John Doe"
      }
    });
    component.update();
    expect(component.state('errorMessage')).toBe('');
  });

  it('should validate input on input and set state.errorMessage if holder name is empty', () => {
    input.prop('onInput')({
      target: {
        value: ""
      }
    });
    component.update();
    expect(component.state('errorMessage')).not.toBe('');
  });

  it('should validate input on input and set state.errorMessage if holder name length is less then 2 symbols', () => {
    input.prop('onInput')({
      target: {
        value: "a"
      }
    });
    component.update();
    expect(component.state('errorMessage')).not.toBe('');
  });

  it('should validate input on input and set state.errorMessage if holder name length is more then 100 symbols', () => {
    input.prop('onInput')({
      target: {
        value: new Array(102).join('a')
      }
    });
    component.update();
    expect(component.state('errorMessage')).not.toBe('');
  });

  it('should validate input on input and set state.errorMessage if holder name does not match validation pattern', () => {
    input.prop('onInput')({
      target: {
        value: "%!@#"
      }
    });
    component.update();
    expect(component.state('errorMessage')).not.toBe('');
  });

  it('should call updateErrorsMock with true error value if validation fails on user input', () => {
    input.prop('onInput')({
      target: {
        value: "%!@#"
      }
    });
    component.update();
    expect(updateErrorsMock.mock.calls[0][0]).toEqual({ key: 'card_holder', value: true });
  });

  it('should call updateErrorsMock with true error value if validation pass on user input', () => {
    input.prop('onInput')({
      target: {
        value: "John Doe"
      }
    });
    component.update();
    expect(updateErrorsMock.mock.calls[0][0]).toEqual({ key: 'card_holder', value: false });
  });

  it('should not display clear button if input value is empty', () => {
    const clearButton = component.find('.card-holder-input__clear-button');
    expect(clearButton.length).toBe(0);
  });

  it('should display clear button if input value is not empty', () => {
    component.setProps({ value: "John Doe" });
    component.update();
    const clearButton = component.find('.card-holder-input__clear-button');
    expect(clearButton.length).not.toBe(0);
  });

  it('should clear input on clear button click', () => {
    component.setProps({ value: "John Doe" });
    component.update();
    const clearButton = component.find('.card-holder-input__clear-button');
    clearButton.prop('onClick')();
    expect(updateFieldsMock.mock.calls[0][0]).toEqual({ card_holder: '' });
  });

  it('should set state.errorDisabled to false on clear button click', () => {
    component.setProps({ value: "John Doe" });
    component.update();
    const clearButton = component.find('.card-holder-input__clear-button');
    clearButton.prop('onClick')();
    component.update();
    expect(component.state('errorDisabled')).toBe(false);
  });

  it('should set state.errorMessage with error for empty field on clear button click', () => {
    component.setProps({ value: "John Doe" });
    component.update();
    const clearButton = component.find('.card-holder-input__clear-button');
    clearButton.prop('onClick')();
    component.update();
    expect(component.state('errorMessage')).not.toBe('');
  });

  it('should call updateErrorsMock method from props and pass proper object if validation fails on user input or clear button', () => {
    component.setProps({ value: "John Doe" });
    component.update();
    const clearButton = component.find('.card-holder-input__clear-button');
    clearButton.prop('onClick')();
    component.update();
    expect(updateErrorsMock.mock.calls[0][0]).toEqual({ key: 'card_holder', value: true });
  });

  it('should change state.errorDisabled to false on blur', () => {
    input.prop('onBlur')();
    component.update();
    expect(component.state('errorDisabled')).toBe(false);
  });

  it('should change trim input value on blur', () => {
    component.setProps({ value: "John Doe " });
    component.update();
    input.prop('onBlur')();
    expect(updateFieldsMock.mock.calls[0][0]).toEqual({ card_holder: 'John Doe' });
  });
});
