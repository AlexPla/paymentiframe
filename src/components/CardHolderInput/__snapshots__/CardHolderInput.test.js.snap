// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component CardHolderInput: should match snapshot 1`] = `
<div
  className="card-holder-input__input-group"
>
  <input
    autoComplete="on"
    className="card-holder-input__input false false"
    id="cardHolderInput"
    onBlur={[Function]}
    onInput={[Function]}
    type="text"
    value=""
  />
  <label
    className="card-holder-input__label"
    htmlFor="cardHolderInput"
  >
    Titular
  </label>
  <svg
    className="card-holder-input__icon"
    height="20px"
    preserveAspectRatio="xMidYMid meet"
    version="1.1"
    viewBox="0 0 150 150"
    width="20px"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M95.540625,75.225 C108.1875,68.296875 116.75625,55.21875 116.75625,40.21875 C116.75625,18.046875 98.025,0 75,0 C51.975,0 33.24375,18.046875 33.24375,40.21875 C33.24375,55.21875 41.8125,68.296875 54.459375,75.225 C23.071875,83.85 0,111.61875 0,144.525 C0,144.890625 0.028125,145.2375 0.084375,145.95 C0.084375,148.190625 1.89375,150 4.134375,150 L145.875,150 C148.10625,150 149.953125,147.825 149.953125,145.584375 C149.971875,145.2375 150,144.890625 150,144.525 C150,111.61875 126.928125,83.85 95.540625,75.225 L95.540625,75.225 Z M41.353125,40.21875 C41.353125,22.51875 56.446875,8.109375 75,8.109375 C93.553125,8.109375 108.646875,22.51875 108.646875,40.21875 C108.646875,57.946875 93.553125,72.365625 75,72.365625 C56.446875,72.365625 41.353125,57.946875 41.353125,40.21875 L41.353125,40.21875 Z M8.165625,141.890625 C9.609375,107.8125 39.0375,80.5125 75,80.5125 C110.840625,80.5125 140.19375,107.64375 141.796875,141.88125 L8.165625,141.88125 L8.165625,141.890625 Z"
    />
  </svg>
</div>
`;
